components/nimbus/src/tests/client/test_http_client.rs

```diff
@@ -2,14 +2,10 @@
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use crate::client::{http_client::*, SettingsClient};
-use crate::{
-    error::NimbusError, Branch, BucketConfig, Experiment, FeatureConfig, RandomizationUnit,
-    RemoteSettingsConfig, SCHEMA_VERSION,
-};
+use crate::client::SettingsClient;
+use crate::{Branch, BucketConfig, Experiment, FeatureConfig, RandomizationUnit, SCHEMA_VERSION};
 use mockito::mock;
-use std::cell::Cell;
-use std::time::{Duration, Instant};
+use rs_client::{Client, ClientConfig};
 
 #[test]
 fn test_fetch_experiments_from_schema() {
```

Comment:
a possible followup: 

I paused here and I'm not sure I have a good answer to my own following question 😬

What **should** this test actually test? I feel a little weird that we are actually sending an HTTP request here now that the client itself is down in `rs-client`.

This should be a nice-to-have follow up but there might be an opportunity here to use trait mocking to mock the `SettingsClient`, see how [`FxAClient`](https://github.com/mozilla/application-services/blob/bd5ff1a29747702cbc95be998903be206c24506a/components/fxa-client/src/internal/http_client.rs#L55-L56) is mocked, and then [the tests depend on a generated type that mocks it](https://github.com/mozilla/application-services/blob/bd5ff1a29747702cbc95be998903be206c24506a/components/fxa-client/src/internal/profile.rs#L124-L141). 

That way we remove both the `viaduct_reqwest` and the mockito dependencies for nimbus! but all this is test code, so later follow up 😄

Reply:
I do like this idea, but I would love some input from the Nimbus team on whether they find a more integration-y test useful to keep around before I make changes. Maybe @jeddai can weigh in?

Reply:
Yeah for the sake of this PR being a "move the rs client down to its own crate" I'm sure we can land this as is, good follow up though to play around with trait mocking

Reply:
So my thought on this is that the structure of this test – calling `get` on the client and validating the type and value of the response – should move to the `rs_client` module. For this new test, the library making the http call and the function's response value are under test. 

As for this test, while it _can_ remain here, what I think this is actually testing now is the `fetch_experiments` and `parse_experiments` methods in `http_client`. I think this should become two different sets of tests:
1. A test on `fetch_experiments` to validate that it a) calls `self.get()`, and b) passes the `text` from that into `parse_experiments`. 
    * This could very likely be done with mocking, though I'll admit I've not had to do much mock testing in Rust yet so I'm not sure off the top of my head how to do that.
2. Tests covering the potential branches of `parse_experiments`. Admittedly this seems like tech debt, and I don't think it should be required for your PR, but the only cases I've found of `parse_experiments` being tested cover the `Ok` case.

Reply:
@tarikeshaq After looking into this, I have some more questions about what we'd like to do here. I'm happy to address this in a follow-up or now in case it has any impact on the API of `rs_client`. Also happy to include an additional Nimbus test to capture a failure branch for `parse_experiments`. The more Rust practice the better 🙂 

I'm not sure that using trait mocking similar to `FxAClient` as mentioned will accomplish the goals stated in @jeddai's post directly above this. That kind of mocking seems to be useful when you have a trait that is being used by a separate type under test, but it seems in both of the cases we really want to test a specific implementation of the trait itself.  

For example:
```
#[cfg_attr(test, mockiato::mockable)]
pub(crate) trait SettingsClient {
    fn get_experiments_metadata(&self) -> Result<String>;
    fn fetch_experiments(&self) -> Result<Vec<Experiment>>;
}

#[test]
fn test_fetch_experiments_from_schema() {
    let client = SettingsClientMock::new()
    client
        .expect_fetch_experiments(...)
        .returns_once(...);
    let resp = client.fetch_experiments().unwrap();
    assert(resp...);
}
```
seems to only really be testing the mocking framework itself.  

I'm curious if, instead, there is a way to mock the type the trait is implemented _for_, in this case `rs_client::Client`. Is it possible to expose that base type as mockable from the support component? Or does this indicate that we might want to change the underlying design, like changing `Client` to be a trait?

For example, I think we want something closer to:
```
fn test_parse_fetched_experiments() {
    let client = ClientMock::new()
    client
        .expect_get_records()
        .returns_once(<response>);

    let resp = client.fetch_experiments().unwrap();
    assert(<parsed response>);
}

fn test_fail_parse_fetched_experiments() {
    ...
}
```

The other option would be to move forward with mocking `SettingsClient` in a higher-level test, but then I think we're missing out on covering the cases above.

Reply:
Thanks a ton for looking deeply into this @MatthewTighe 🚀 

> That kind of mocking seems to be useful when you have a trait that is being used by a separate type under test, but it seems in both of the cases we really want to test a specific implementation of the trait itself.

That is exactly right - if we are testing behaviour specific to the HTTP client, which from @jeddai 's responses is what we want here, then mocking the trait won't help us!


> I'm curious if, instead, there is a way to mock the type the trait is implemented for, in this case rs_client::Client

Unfortunately Rust doesn't let us consume test-only types in other crates :( So even if we had defined `rs_client::Client` as a trait first, its mock won't be consumable here in Nimbus.


**Let's open a follow-up ticket for this case**. I think it's a small enough detail that we can land with the test as-is

(the only win in making this better is getting rid of a test-only dependency!)


FWIW I believe a way forward here could include:
- Exposing a `ClientTrait` trait from `rs_client`, and having `rs_client::Client` implement it
- The Nimbus Http Client own either a `Box<dyn ClientTrait>` or be generic on a type that implements `ClientTrait`
- In the test code, we can create our own TestClient type that implements `ClientTrait` and returns whatever we want it to return, then we can validate the behavior of the nimbus client.


That all adds complexity though and I'm not exactly sure is the best way forward, so a follow-up is probably best!

Reply:
I agree on the follow-up ticket. I'm going to do a quick scan through the rest of the changes and will approve after that!

Reply:
Just to close the loop, filed https://github.com/mozilla/application-services/issues/5394 as a follow-up and will be working on it shortly.